File: release/release.pl
#!/usr/bin/perl
# Release ../trunk/tvrenamer.pl
# 1) Update $version string with latest version in script header and current date
# 2) Compile windows executable from a DOS shell
# 3) Upload

qx{/cygdrive/c/Perl/bin/perl.exe win32_release.pl};
if(@ARGV[0] eq "beta")
{
    print "Renaming to betas...\n";
    rename("tvrenamer.pl", "tvrenamer.beta.pl");
    rename("tvrenamer.exe", "tvrenamer.beta.exe");
    rename("tvrenamer.noansi.exe", "tvrenamer.noansi.beta.exe");
    $script = "tvrenamer.beta.pl";
    $binary = "tvrenamer.beta.exe";
    $64bit_binary = "tvrenamer.noansi.beta.exe";
}
else
{
    $script = "tvrenamer.pl";
    $binary = "tvrenamer.exe";
    $64bit_binary = "tvrenamer.noansi.exe";
}
print "Uploading...";
system("scp $script $binary $64bit_binary robmeerman.co.uk:public_html/downloads");
print "Done.\n";

# Tag if upload was successful (and the environment doesn't have a NOTAG
# variable)
# $?'s high byte is the return code (low byte is the signal it died with, if
# any)
$retcode = $? >> 8;
if(0 == $retcode and @ENV{NOTAG} eq "" and @ARGV[0] ne "beta")
{
    print "Tagging this release...\n";
    system('./tag.sh');
    print "Done\n";
}
system("cmd /C 'pause'");

File: release/release.py
#!/usr/bin/env python
# Update $version in tvrenamer.pl based on current date and highest version 
# number in change-log comments

import os
import re
import time
import sys

beta_release = False
if "--beta" in sys.argv[1:]:
    beta_release = True


p = os.popen("bash -c 'pushd .. >/dev/null; git ls-files -m trunk/tvrenamer.pl; popd >/dev/null'")
output = p.read()
p.close()
if len(output) > 0:
    print "tvrenamer.pl has uncommitted modifications!"
    exit(1);

with file("../trunk/tvrenamer.pl", "r") as fh:
    script = fh.read()

versions = []
for match in re.finditer("#\s*v(\d+)\.(\d+)\s", script):
    versions.append( (int(match.group(1)), int(match.group(2))) )
versions = sorted(versions)
release = versions[-1]
release = [ str(x) for x in release ]
release = ".".join(release)
release = "v" + release

today = time.strftime("%d %B %Y")

print "Updating --version info to %s (%s)" % (release, today)

# Note that the replacement string is processed by re.subn, and so must be 
# escaped *in addition to* being a 'raw' string
script, changes = re.subn(
    r'my \$version \= \"TV Series Renamer \d+\.\d+\\nReleased [^"]+\\n\"\; \# \{\{\{',
    r"""my $version = "TV Series Renamer %s\\nReleased %s\\n"; # {{{""" % (
            release,
            today,
            ),
    script
    )

if changes == 0:
    print "Release script was unable to update --version info, please fix it!"
    exit(2)

if changes > 1:
    print "Release script got confused by multiple '$version' lines in tvrenamer.pl, please update it."
    exit(3)

with file("../trunk/tvrenamer.pl", "w") as fh:
    fh.write(script)

if beta_release:
    dst = "tvrenamer.beta.pl"
else:
    dst = "tvrenamer.pl"

print "Uploading %s..." % dst
retval = os.system("rsync -P ../trunk/tvrenamer.pl robmeerman.co.uk:public_html/downloads/%s" % dst)
print "Done"

if retval == 0:
    print "Upload successful, tagging release"
    os.system("git add -u ../trunk/tvrenamer.pl")
    if beta_release:
        os.system("git commit -m 'Beta %s'" % release)
        os.system("git tag -f %s" % release)
    else:
        os.system("git commit -m 'Release %s'" % release)
        os.system("git tag %s" % release)
